src/main/java/bio/terra/janitor/service/cleanup/FlightManager.java

```diff
@@ -118,5 +120,95 @@ private boolean submitToStairway(String flightId, TrackedResource resource) {
+    try {
+      flightState = stairway.getFlightState(flightId);
+    } catch (DatabaseOperationException | InterruptedException e) {
+      logger.error(String.format("Error getting state of finishing flight [%s]", flightId), e);
+      return false;
+    }
+    TrackedResourceState endCleaningState;
+    if (flightState.getFlightStatus().equals(FlightStatus.SUCCESS)) {
+      endCleaningState = TrackedResourceState.DONE;
+    } else if (flightState.getFlightStatus().equals(FlightStatus.ERROR)) {
+      endCleaningState = TrackedResourceState.ERROR;
+    } else {
+      // The flight hasn't finished or has finished fatally. Let Stairway keep working or the fatal
+      // monitor handle this respectively.
+      return false;
+    }
+    try {
+      updateFinishedCleanupState(
+          resourceAndFlight.trackedResource().trackedResourceId(), flightId, endCleaningState);
+    } catch (UnexpectedCleanupState unexpectedCleanupState) {
+      logger.error(
+          String.format("Error finishing flight changing cleanup state. [%s]", flightId),
+          unexpectedCleanupState);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Transactionally update the TrackedResourceState and CleanupFlightState for the finishing
+   * flight. Throws an exception if there is an unexpected state to rollback the transaction.
+   */
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  private void updateFinishedCleanupState(
+      TrackedResourceId trackedResourceId, String flightId, TrackedResourceState endCleaningState)
+      throws UnexpectedCleanupState {
+    // Retrieve the TrackedResource within this transaction to ensure no one else is abandoning or
+    // duplicating it while we finish it.
+    Optional<TrackedResource> resource = janitorDao.retrieveTrackedResource(trackedResourceId);
+    if (!resource.isPresent()) {
+      throw new UnexpectedCleanupState(
+          String.format("Unable to find tracked_resource with id [%s]", trackedResourceId));
+    }
+    TrackedResourceState resourceState = resource.get().trackedResourceState();
+    if (resourceState.equals(TrackedResourceState.CLEANING)) {
+      janitorDao.updateResourceState(trackedResourceId, endCleaningState);
+      // We assume no one else is modifying the CleanupFlightState while we do this.
+    } else if (!resourceState.equals(TrackedResourceState.ABANDONED)
+        && !resourceState.equals(TrackedResourceState.DUPLICATED)) {
+      // The resource should not have moved from CLEANING to any other state while there was a
+      // flight working on it.
+      throw new UnexpectedCleanupState(
+          String.format("Unexpected TrackedResourceState: %s", resourceState));
+    }
+    janitorDao.updateFlightState(flightId, CleanupFlightState.FINISHED);
+  }
+
+  /** Exception for unexpected resource state when finishing a cleanup flight. */
+  private static class UnexpectedCleanupState extends Exception {
```

Comment:
UnexpectedCleanupStateException? And do we need to make this has its own class?

Reply:
So we need to throw something in order to get the rollback. I thought it made sense for it to be its own class so that we can catch only violations to our own logical invariants and log them accordingly. Then we can handle them separately from unexpected exceptions. But those are both forms of unexpected, so we could not special case them and throw something like `IllegalStateException`. WDYT?